adaptive/learner/balancing_learner.py

```diff
@@ -111,38 +111,57 @@ def strategy(self, strategy):
             raise ValueError(
                 'Only strategy="loss_improvements", strategy="loss", or'
                 ' strategy="npoints" is implemented.')
+        self._points = {}  # reset the cache
 
     def _ask_and_tell_based_on_loss_improvements(self, n):
-        points = []
-        loss_improvements = []
+        chosen_points = []
```

Comment:
```python
selected = []  # tuples (learner_index, point) 
```

Reply:
Overall we should make a clear distinction between `point` and `(learner_index, point)`. Conceptually these are the same (the latter is a point from a balancing learner perspective). This makes them easy to confuse, and therefore a clear name separation is very useful.

Reply:
Maybe we could make `selected` like:

`selected = []  # tuples ((learner_index, point), loss_improvement) `

Reply:
Indeed, however shouldn't it be flattened as well then?

Reply:
Yes, like I did in the last commit?

Reply:
Yes, but what about flattening this data structure as well?

Reply:
NB: PEP8 suggests a double space before a trailing comment.

Reply:
Fixed!

Reply:
Why not also flatten, so `(learner_index, point, loss_improvement)`?

Reply:
Disregard that; it's the format the runner expects.